packages/aws-cdk-lib/aws-stepfunctions/lib/state-machine.ts

```diff
@@ -517,18 +517,8 @@ export class StateMachine extends StateMachineBase {
       if (props?.enableEncryptedLogging && props?.logs) {
         props.kmsKey.addToResourcePolicy(new iam.PolicyStatement({
           resources: ['*'],
-          actions: ['kms:Encrypt*', 'kms:Decrypt*', 'kms:ReEncrypt*', 'kms:GenerateDataKey*', 'kms:DescribeKey*'],
+          actions: ['kms:Encrypt*', 'kms:Decrypt*', 'kms:ReEncrypt*', 'kms:GenerateDataKey*', 'kms:Describe*'],
           principals: [new iam.ServicePrincipal(`logs.${Stack.of(this).region}.amazonaws.com`)],
-          conditions: {
-            ArnEquals: {
-              'kms:EncryptionContext:aws:logs:arn': Stack.of(this).formatArn({
-                service: 'logs',
-                resource: 'log-group',
-                sep: ':',
-                resourceName: props.logs.destination.logGroupName,
-              }),
-            },
-          },
```

Comment:
Adding the encryption context here was causing a circular dependency issue when trying to deploy a CDK stack using a KMS key with a StateMachine and CWL Log Group: 

```
export class HelloCdkStack extends cdk.Stack {
  testStateMachine: cdk.aws_stepfunctions.StateMachine;
  testActivity: cdk.aws_stepfunctions.Activity;

  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const key = new kms.Key(this, 'Key used for encryption');

    const activity = new Activity(this, 'TestActivity', {
      activityName: 'TestActivty',
      kmsKey: key,
      kmsDataKeyReusePeriodSeconds: cdk.Duration.seconds(75),
    });

    const logGroup = new logs.LogGroup(this, 'MyLogGroup', {
      logGroupName: '/aws/vendedlogs/states/MyLogGroup',
      encryptionKey: key,
    });

    const stateMachine = new StateMachine(this, 'StateMachineWithCMKWithCWLEncryption', {
      stateMachineName: 'StateMachineWithCMKWithCWLEncryption',
      definitionBody: DefinitionBody.fromChainable(Chain.start(new Pass(this, 'Pass'))),
      stateMachineType: StateMachineType.STANDARD,
      kmsKey: key,
      kmsDataKeyReusePeriodSeconds: cdk.Duration.seconds(75),
      enableEncryptedLogging: true,
      logs: {
        destination: logGroup,
        level: LogLevel.FATAL,
        includeExecutionData: false,
      },
    });
  }

```

I was seeing the following error:

```
❌ Deployment failed: Error [ValidationError]: Circular dependency between resources: [MyLogGroup5C0DAD85, Key961B73FD, StateMachineWithCMKWithCWLEncryptionE07EB791]
    at Request.extractError (/Users/saravaid/Desktop/aws-cdk/packages/aws-cdk/node_modules/aws-sdk/lib/protocol/query.js:50:29)
    at Request.callListeners (/Users/saravaid/Desktop/aws-cdk/packages/aws-cdk/node_modules/aws-sdk/lib/sequential_executor.js:106:20)
    at Request.emit (/Users/saravaid/Desktop/aws-cdk/packages/aws-cdk/node_modules/aws-sdk/lib/sequential_executor.js:78:10)
    at Request.emit (/Users/saravaid/Desktop/aws-cdk/packages/aws-cdk/node_modules/aws-sdk/lib/request.js:686:14)
    at Request.transition (/Users/saravaid/Desktop/aws-cdk/packages/aws-cdk/node_modules/aws-sdk/lib/request.js:22:10)
    at AcceptorStateMachine.runTo (/Users/saravaid/Desktop/aws-cdk/packages/aws-cdk/node_modules/aws-sdk/lib/state_machine.js:14:12)
    at /Users/saravaid/Desktop/aws-cdk/packages/aws-cdk/node_modules/aws-sdk/lib/state_machine.js:26:10
    at Request.<anonymous> (/Users/saravaid/Desktop/aws-cdk/packages/aws-cdk/node_modules/aws-sdk/lib/request.js:38:9)
    at Request.<anonymous> (/Users/saravaid/Desktop/aws-cdk/packages/aws-cdk/node_modules/aws-sdk/lib/request.js:688:12)
    at Request.callListeners (/Users/saravaid/Desktop/aws-cdk/packages/aws-cdk/node_modules/aws-sdk/lib/sequential_executor.js:116:18) {
  code: 'ValidationError',
  time: 2024-08-10T07:39:07.240Z,
  requestId: '23d39dcb-71ab-4267-8ddc-57980fa64a96',
  statusCode: 400,
  retryable: false,
  retryDelay: 680.4362149103878
}
```

The issues lies in this check here where we use a condition which specifies an encryption context:

```
if (props?.enableEncryptedLogging && props?.logs) {
        props.kmsKey.addToResourcePolicy(new iam.PolicyStatement({
          resources: ['*'],
          actions: ['kms:Encrypt*', 'kms:Decrypt*', 'kms:ReEncrypt*', 'kms:GenerateDataKey*', 'kms:Describe*'],
          principals: [new iam.ServicePrincipal(`logs.${Stack.of(this).region}.amazonaws.com`)],
          conditions: {
            ArnEquals: {
              'kms:EncryptionContext:aws:logs:arn': Stack.of(this).formatArn({
                service: 'logs',
                resource: 'log-group',
                sep: ':',
                resourceName: props.logs.destination.logGroupName ----> this is causing the circular dependency issue,
              }),
            },
          },
        }));

```
I think a way to mitigate this issue and still make it easier for customers is removing the encryption context, by doing so the [additional authenticated data](https://docs.aws.amazon.com/crypto/latest/userguide/cryptography-concepts.html#term-aad) will not be bound to the encrypt and decrypt operations.

Reply:
I'm going to resolve this, it's better to use two separate keys since each key can have its own responsibilities and own permission sets. With that, we won't be running into a circular dependency issue since we'd be using two keys and removing the dependency between the single key and the log group